core/tracking.go

```diff
@@ -0,0 +1,44 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"context"
+
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+)
+
+// WithTracking wraps component input functions to support tracking of core components.
+func WithTracking(tracker Tracker) WireOption {
```

Comment:
component input wrapping order is very important, what should the order be between retryer, tracing and tracking?

Reply:
after tracing and before retryer:
```
core.Wire(sched, fetch, cons, dutyDB, vapi,
		parSigDB, parSigEx, sigAgg, aggSigDB, broadcaster,
		core.WithTracing(),
		core.WithTracking(tracker),
		core.WithAsyncRetry(retryer),
	)
```

Reply:
retryer is async so it always return nil error immediately, so it has to "after" it. But retryer might retry, so it will result in multiple events being added to tracker, one for each retry.

Reply:
maybe we can dedup events added to tracker by component/step per duty?

Reply:
not sure how that is going to work?